cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -327,6 +330,15 @@ public void performAction(HttpRequest request, HttpResponder responder,
                             @PathParam("type") String type,
                             @PathParam("id") String id,
                             @PathParam("action") String action) {
+    // If the app is deployed and it is an Application Template, then don't allow any action.
+    // Operations are only allowed through Adapter Lifecycle management.
+    if (store.getApplication(Id.Application.from(namespaceId, appId)) != null &&
```

Comment:
Hmm for example, say the user wants to deploy an app which matches with a apptemplate name. Should we not allow that even though no adapter was created? Though the user won't be able to create an adapter of that type after that (until he deletes his app). Thoughts?